fog/view/server/src/fog_view_router_request_handler.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{
+    error::{router_server_err_to_rpc_status, RouterServerError},
+    fog_view_shard_responses_processor,
+};
+
+use futures::{future::try_join_all, SinkExt, TryStreamExt};
+use grpcio::{ChannelBuilder, DuplexSink, RequestStream, RpcStatus, WriteFlags};
+use mc_attest_api::attest;
+use mc_common::{logger::Logger, ResponderId};
+use mc_fog_api::{
+    view::{
+        FogViewRouterRequest, FogViewRouterResponse, MultiViewStoreQueryRequest,
+        MultiViewStoreQueryResponse,
+    },
+    view_grpc::FogViewApiClient,
+};
+use mc_fog_uri::FogViewUri;
+use mc_fog_view_enclave_api::ViewEnclaveProxy;
+use mc_util_grpc::{rpc_invalid_arg_error, ConnectionUriGrpcioChannel};
+use std::{str::FromStr, sync::Arc};
+
+const RETRY_COUNT: usize = 3;
+
+/// Handles a series of requests sent by the Fog Router client.
+pub async fn handle_requests<E>(
+    shard_clients: Vec<Arc<FogViewApiClient>>,
+    enclave: E,
+    mut requests: RequestStream<FogViewRouterRequest>,
+    mut responses: DuplexSink<FogViewRouterResponse>,
+    logger: Logger,
+) -> Result<(), grpcio::Error>
+where
+    E: ViewEnclaveProxy,
+{
+    while let Some(request) = requests.try_next().await? {
+        let result = handle_request(
+            request,
+            shard_clients.clone(),
+            enclave.clone(),
+            logger.clone(),
+        )
+        .await;
+        match result {
+            Ok(response) => responses.send((response, WriteFlags::default())).await?,
```

Comment:
I'm curious as to what a `feed` implementation would look like? Would we put logic here that would determine when we should call `flush`? Or would we rely on the client to call flush when they want? It seems like the `flush` logic might make things complicated, but I may be misunderstanding the API.

From the `feed` API Docs:
`Unlike send, the returned future does not flush the sink. It is the caller’s responsibility to ensure all pending items are processed, which can be done via flush or close.`

Reply:
I think it would be something along the lines of
```rust
    responses.flush().await?.close().await?;
```
one possible downside is if you do error out in the flush() you don't know which response was the issue, not sure if you know that now.
it's probably premature optimization at this point

Reply:
Thinking about this more, `send` is probably the way to go.
My understanding is that requests will keep coming on the stream until the client disconnects, this means that potentially no responses come out until a client disconnects, but it seems possible a client stays connected indefinitely.
it would be possible to use feed and some timed flush, but that seems like a bunch of logic and overhead without justification at this time.